### DISCUSSION TITLE
External transaction with asyncio

### DISCUSSION THREAD

**Discussion by e-kondr01** [sqlalchemy/sqlalchemy#10857] [Usage Questions]
I'm writing unit tests with `pytest` and async `sqlalchemy`.  The goal is to rollback all DB changes made within a test function, even though calls to `session.commit()` may be present. I used the following recipe:

https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites

And adapted it for `asyncio`. This is the code that works correctly for me:

```python
engine = create_async_engine(...)

@pytest.fixture()
async def session() -> AsyncGenerator[AsyncSession, None]:
    async with engine.connect() as connection:
        async with connection.begin() as transaction:
            async_session = AsyncSession(
                bind=connection, join_transaction_mode="create_savepoint"
            )

            yield async_session

            await transaction.rollback()
    await engine.dispose()
```

So, I had to add entering of async contexts and a call to `engine.dispose()` (to aviod `RuntimeError: Task ... attached to a different loop` as per https://github.com/sqlalchemy/sqlalchemy/issues/5626#issuecomment-1062470136).

Would it be the correct way to implement this functionality with `asyncio` `sqlalchemy`? The tests behave as desired as far as I can see, for example these tests are successful:

```python
async def test_db_change(session: AsyncSession):
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 0
    session.add(Profile(name="test", surname="test"))
    await session.commit()
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 1


async def test_db_change2(session: AsyncSession):
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 0
```
Output:
```shell
============================= test session starts ==============================
platform linux -- Python 3.12.1, pytest-7.4.4, pluggy-1.3.0
rootdir: /app
configfile: pytest.ini
plugins: anyio-3.7.1, asyncio-0.23.3
asyncio: mode=Mode.AUTO
collected 3 items

app/tests/test_demo.py ...                                               [100%]

=============================== warnings summary ===============================
../usr/local/lib/python3.12/site-packages/dateutil/tz/tz.py:37
  /usr/local/lib/python3.12/site-packages/dateutil/tz/tz.py:37: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    EPOCH = datetime.datetime.utcfromtimestamp(0)

../usr/local/lib/python3.12/site-packages/aio_pika/abc.py:41
  /usr/local/lib/python3.12/site-packages/aio_pika/abc.py:41: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    ZERO_TIME = datetime.utcfromtimestamp(0)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 3 passed, 2 warnings in 0.12s =========================
```

**Comment by zzzeek** [comment]
that seems perfect to me, in fact I think we should have this documented in the sqlalchemy asyncio docs and "seealso" linked from the session docs you mention.   Can you provide the complete example and we can put it up?

**Reply by zzzeek** [reply]
is this using pytest-asyncio ?

**Reply by e-kondr01** [reply]
A more detailed example:

```python
from typing import AsyncGenerator
from uuid import UUID, uuid4

import pytest
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

engine = create_async_engine("postgresql+psycopg2://...")


class Profile(DeclarativeBase):
    id: Mapped[UUID] = mapped_column(
        primary_key=True,
        default=uuid4,
        server_default=func.gen_random_uuid(),
    )
    name: Mapped[str]


pytest_plugins = ("pytest_asyncio",)


@pytest.fixture()
async def session() -> AsyncGenerator[AsyncSession, None]:
    async with engine.connect() as connection, connection.begin() as transaction:
        async_session = AsyncSession(
            bind=connection, join_transaction_mode="create_savepoint"
        )

        yield async_session

        await transaction.rollback()
    await engine.dispose()


async def test_create_profile(session: AsyncSession):
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 0

    test_name = "test"
    session.add(Profile(name=test_name))
    await session.commit()

    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 1
    assert existing_profiles[0].name == test_name


async def test_rollbacks_between_functions(session: AsyncSession):
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 0
```

A few things to note:

1. This requires the following `pytest.ini` file:
```
[pytest]
asyncio_mode=auto
```
Alternatively, test functions can be decorated with `@pytest.mark.asyncio`

2. I'm managing migrations with `alembic`, so there is no creation of tables with `sqlalchemy`

**Reply by e-kondr01** [reply]
Installed testing packages:
```
pytest==7.4.4
pytest-asyncio==0.23.3
```

**Reply by CaselIT** [reply]
Thanks

The only unfortunate thing is that pytests-asyncio seems to create a new loop for every test, requiring the use of `engine.dispise`. Do you know if it's possible telling it to reuse the same loop on all tests?

**Reply by e-kondr01** [reply]
Actually, the main reason for this thread was that I wondered if calling `engine.dispose()` could be avoided.

I went on to read more about event loops in `pytest-asyncio` (https://pytest-asyncio.readthedocs.io/en/latest/concepts.html#asyncio-event-loops).

So with `pytest-asyncio` it is possible to use one event loop in several test functions, if they have the same scope. The complication arises with the scope of the fixture.

If I mark the session fixture with the same scope as test functions (for example, `scope=module`), then there are no errors, but it functions incorrectly, as one session is used for all of the functions within the scope, and there are no rollbacks between functions.

If I mark test functions with the same scope higher than function, and leave fixture's scope as default (function scope), than `pytest-asyncio` raises `pytest_asyncio.plugin.MultipleEventLoopsRequestedError`.  

However, I had more luck using `pytest-anyio`. Its docs say "if even one async fixture, scoped higher than function, is shared across all tests, only one test runner will be created during the test session" (https://anyio.readthedocs.io/en/stable/testing.html#technical-details)
So I came up with solution that doesn't require disposing of engine because of using one event loop, which I am posting below.

**Reply by e-kondr01** [reply]
```python
from typing import AsyncGenerator
from uuid import UUID, uuid4

import pytest
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncConnection, AsyncSession, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

pytestmark = pytest.mark.anyio

engine = create_async_engine("postgresql+psycopg2://...")


class Profile(DeclarativeBase):
    id: Mapped[UUID] = mapped_column(
        primary_key=True,
        default=uuid4,
        server_default=func.gen_random_uuid(),
    )
    name: Mapped[str]


@pytest.fixture(scope="session")
def anyio_backend():
    return "asyncio"


@pytest.fixture(scope="session")
async def connection(anyio_backend):
    async with engine.connect() as connection:
        yield connection


@pytest.fixture()
async def session(connection: AsyncConnection) -> AsyncGenerator[AsyncSession, None]:
    async with connection.begin() as transaction:
        async_session = AsyncSession(
            bind=connection, join_transaction_mode="create_savepoint"
        )

        yield async_session

        await transaction.rollback()


async def test_create_profile(session: AsyncSession):
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 0

    test_name = "test"
    session.add(Profile(name=test_name))
    await session.commit()

    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 1
    assert existing_profiles[0].name == test_name


async def test_rollbacks_between_functions(session: AsyncSession):
    existing_profiles = (await session.execute(select(Profile))).scalars().all()
    assert len(existing_profiles) == 0
```

**Reply by zzzeek** [reply]
@CaselIT we run our whole test suite in one event loop?

**Reply by e-kondr01** [reply]
This version does not use `pytest-asyncio` and shares one `AsyncConnection` between all test functions.

The only thing I am concerned about is whether sharing a connection will prevent these tests from running in parallel (haven't worked with running tests with `pytest` in parallel, but it may be required to speed up test runs).

**Reply by zzzeek** [reply]
there's no "parallelism" in Python other than multiprocessing which is pytest-xdist, so if you use that you just need to ensure that every worker process starts with an "empty" engine (e.g. pool is disposed or otherwise not started).

there's also the idea to just use `NullPool` and then we dont need to dispose anything.

**Reply by CaselIT** [reply]
> @CaselIT we run our whole test suite in one event loop?

yes, the test now use a single [Runner](https://docs.python.org/3/library/asyncio-runner.html#asyncio.Runner) that uses a single loop.

I honestly don't understand why pytest-asyncio decided to change how it works and now creates a new loop for each test function (in the default case), since normally programs use a single loop anyway.

**Reply by e-kondr01** [reply]
Did a quick test with `pytest-xdist`, no unexpected errors when running with multiple processes, all my tests pass. So the latest example with `pytest-anyio` seems to be OK to run tests not disposing of `AsyncEngine`
